plutus-ledger/src/Ledger/Tx/CardanoAPI.hs

```diff
@@ -627,32 +633,52 @@ toCardanoMintValue redeemers value mps =
         <*> (C.BuildTxWith . Map.fromList <$> traverse (\(idx, mp) -> (,) <$> (toCardanoPolicyId . P.mintingPolicyHash) mp <*> toCardanoMintWitness redeemers idx mp) (Prelude.zip [0..] $ Set.toList mps))
 
 fromCardanoValue :: C.Value -> P.Value
-fromCardanoValue (C.valueToList -> list) = foldMap toValue list
-    where
-        toValue (C.AdaAssetId, C.Quantity q) = Ada.lovelaceValueOf q
-        toValue (C.AssetId policyId assetName, C.Quantity q)
-            = Value.singleton (Value.mpsSymbol $ fromCardanoPolicyId policyId) (fromCardanoAssetName assetName) q
+fromCardanoValue (C.valueToList -> list) =
+    foldMap fromSingleton list
+  where
+    fromSingleton (fromCardanoAssetId -> Value.AssetClass (cs, tn), C.Quantity q) =
```

Comment:
I think this can be simplified with `assetClassValue` too?

Reply:
You are right. The naming of those two functions seems to let me reach `singleton` first because it is the underlying concept I am looking for. Thanks.